src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -746,6 +747,19 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
                 },
                 moved_lp.ty));
         }
+        if let (Some(CompilerDesugaringKind::ForLoop), Ok(snippet)) = (
```

Comment:
We might want to use a separate `CompilerDesugaringKind` variant to differentiate in errors, something along the lines of `ForLoopHead`, or something, or modify `ForLoop` to carry information on wether the source is completely synthetic or just moving around code that the user has written.